### DISCUSSION TITLE
logstash stops stashing logs at irregular intervals

### DISCUSSION THREAD

**Discussion by ghb1912** [Security-Onion-Solutions/securityonion#2919] [Unsupported Versions]
Hello all --

I'm running a standalone configuration of 2.3.21 which was installed on Ubuntu 18.04, via a network                                              
installation (not your ISO). I made minimal changes after a default installation of Ubuntu.                                                  
                                                                                                                    
The host has an elderly 4C/8T Xeon, 32G memory, and ~3.5T of storage.                                                 
About 3T of that is allocated to /nsm, and no filesystem is overfilling;                                            
I see that the high water mark is managed nicely around 89% on /nsm, and all the                                            
others are >10% used. Memory (according to grafana) isn't in short                                                  
supply; used memory hovers around 18G or so. I don't see any other                                                  
resources constrained. This host is watching a pair of 1G LAN taps on a small network. 
Neither tap is near saturated ("Monitor traffic" in grafana shows ~170Mb/s at its busiest, generally).                                                             
                                                                                                                    
Every 2-4 days or so (irregular intervals), I notice that no new events are appearing in elasticsearch. 
When this happens, I find that logstash appears to have stopped                                                
consuming from redis. Beginning at the time of the last record in elasticsearch,
the redis queue slowly climbs to max memory, then sits                                            
there until I notice the problem. All services in 'so-status' show green/OK, and no                                                  
resource appears constrained - memory remains steady at ~18G usage on                                               
the host, and no filesystem is full. Traffic is at or near typical.

The only obvious error I see during these blockages is this:

    # salt-call state.highstate
    local:
        Data failed to compile:
    ----------
        The function "state.highstate" is running as PID 25203 and was started at 2021, Feb 09 23:18:33.375383 with jid 
    20210209231833375383

I then find if I do so-logstash-restart, events start getting cleared out of redis, and 'salt-call state.highstate' succeeds:

```
# salt-call state.highstate
[WARNING ] State for file: /etc/pki/ca.crt - Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily.
[WARNING ] State for file: /etc/pki/ca.crt - Neither 'source' nor 'contents' nor 'contents_pillar' nor 'contents_grains' was defined, yet 'replace' was set to 'True'. As there is no source to replace the file with, 'replace' has been set to 'False' to avoid reading the file unnecessarily.

[ ... snip ... ]

----------
          ID: lasthighstate
    Function: file.touch
        Name: /opt/so/log/salt/lasthighstate
      Result: True
     Comment: Updated times on file /opt/so/log/salt/lasthighstate
     Started: 00:06:55.143436
    Duration: 1.99 ms
     Changes:   
              ----------
              touched:
                  /opt/so/log/salt/lasthighstate

Summary for local
--------------
Succeeded: 535 (changed=17)
Failed:      0
--------------
Total states run:     535
Total run time:   209.884 s
```

And everything is right as rain, for another day or few.

Interesting logs:

In the most recent occurrence, the last recorded event in elasticsearch was about 2021-0209 0549 UTC. The timestamp of the most recent state file in the most recent directory under /nsm/elasticsearch/nodes/0/indices/ was about 38 minutes prior to that, at 0511. 

I checked events in /opt/so/log/ for logstash, elasticsearch, redis, and salt between and before this time window, and very little stands out (to me) as unusual or different from messages that occur when everything works. The interesting (to my untrained eye) bits:

/opt/so/log/logstash/logstash.log

```
[2021-02-09T05:09:30,069][INFO ][org.elasticsearch.cluster.metadata.MetadataMappingService] [so-zeek-2021.02.09/sXijTBCZR7K6rWfeGUHt1A] update_mapping [_doc]
[2021-02-09T05:11:13,580][INFO ][org.elasticsearch.cluster.metadata.MetadataMappingService] [so-strelka-2021.02.09/ah9OZ28kT-e8-t9ZS96XEA] update_mapping [_doc]
[2021-02-09T05:11:13,604][INFO ][org.elasticsearch.cluster.metadata.MetadataMappingService] [so-strelka-2021.02.09/ah9OZ28kT-e8-t9ZS96XEA] update_mapping [_doc]
```

That doesn't look onerous, but I call it out only because it's an infrequent entry, and occurred at nearly the same time as the the last elasticsearch state file timestamp. 

I also see copious errors of this nature, but they occur constantly even when everything appears to be working, so I can't say they're directly related:

```
[2021-02-09T05:48:57,891][INFO ][logstash.outputs.elasticsearch] retrying failed action with response code: 500 ({"type"=>"illegal_state_exception", "reason"=>"Pipeline processor configured for non-existent pipeline [suricata.]"})
[2021-02-09T05:48:57,891][INFO ][logstash.outputs.elasticsearch] Retrying individual bulk actions that failed or were rejected by the previous bulk request. {:count=>1}
[2021-02-09T05:49:18,070][INFO ][logstash.outputs.elasticsearch] retrying failed action with response code: 500 ({"type"=>"illegal_state_exception", "reason"=>"Pipeline processor configured for non-existent pipeline [suricata.]"})
[2021-02-09T05:49:18,070][INFO ][logstash.outputs.elasticsearch] Retrying individual bulk actions that failed or were rejected by the previous bulk request. {:count=>1}
[2021-02-09T05:49:20,070][INFO ][logstash.outputs.elasticsearch] retrying failed action with response code: 500 ({"type"=>"illegal_state_exception", "reason"=>"Pipeline processor configured for non-existent pipeline [suricata.]"})
[2021-02-09T05:49:20,070][INFO ][logstash.outputs.elasticsearch] Retrying individual bulk actions that failed or were rejected by the previous bulk request. {:count=>1}
```

salt/minion is also full of these, constantly, and I have no idea what query is failing:
```
`2021-02-09 05:18:18,545 [salt.modules.network:2085][ERROR   ][2365] Exception during resolving address: [Errno 2] Host name lookup failure`
```

There are other errors in that log, but they are "routine" and do not seem to coincide with the failure I'm seeing, such as the initial two lines from salt-call above.

I'm unfortunately not well enough acquainted with all the moving parts to know where to poke next. I would be grateful for some hints about where else I should be looking to understand why logstash is apparently wedging periodically.

Thank you!

**Comment by shanekindt** [comment]
Were you able to figure anything out with your issue?  I had a similar issue (manager-search node) and remote sensor would forward logs via filebeat to logstash and they would pile up in redis.  I had a couple errors in the redis logs (/opt/so/log/redis/redis-server.log) and ended up following an error in there to increase my default logstash batch from 125 to 511 then POOF!  all my redis backlogs flooded to elastic...

Now... logstash is pulling from redis but I am getting errors like yours in my logstash.log file where "Pipeline Processor configured for non-existent pipeline".    Restart logstash and good for a while again.

**Reply by ghb1912** [reply]
Nope. I gave up and just scheduled a logstash restart daily in cron.
